sandbox.go

```diff
@@ -1,130 +1,317 @@
+	err = os.Setenv("DOCKER_API_VERSION", "1.35")
+	if err != nil {
+		return nil, err
+	}
+
+	// init a docker api client
+	dockerClient, err := client.NewEnvClient()
+	if err != nil {
+		// this could occur if docker has not been installed or started
+		return nil, err
+	}
+
+	// TODO (cw|4.29.2018) if we are spinning up this sandbox from within another docker
+	// container, we may want to define a bridge network between them (since they will be
+	// sibling containers). I don't know if this is entirely necessary though...
+	// THIS NETWORK SETUP SHOULD ACTUALLY GO IN A HIGHER SCOPE (within the struct which
+	// actually constructs sandboxes). this way we aren't creating and destroying docker
+	// networks all over the place. instead we should check to see if one has been created.
+
+	// define unique network name
+	// networkName := fmt.Sprintf("xaqt.%s", uuid.NewV4().String())
+
+	// setup container bridge network if one doesn't already exist.
+	// _, err = dockerClient.NetworkCreate(
+	// 	context.TODO(),
+	// 	networkName,
+	// 	types.NetworkCreate{},
+	// )
+	// if err != nil {
+	// 	return nil, err
+	// }
+
+	s = &sandbox{
+		ID:       uuid.NewV4().String(),
+		language: l,
+		code:     code,
+		stdin:    stdin,
+		options:  opts,
+		docker:   dockerClient,
+	}
+
+	return s, nil
 }
 
+// runs user code within the sandbox after preparing the execution environment.
+//
 func (s *sandbox) run() (string, error) {
-	s.prepare()
-	return s.execute()
+	var (
+		output string
+		err    error
+	)
+
+	err = s.prepare()
+	if err != nil {
+		return "", err
+	}
+
+	output, err = s.execute()
+	if err != nil {
+		return "", err
+	}
+
+	return output, nil
 }
 
-func (s *sandbox) prepare() {
-	tmpFolder, err := ioutil.TempDir(s.options.folder, "xaqt-")
+// prepares the execution environment and the sandbox docker container.
+//
+func (s *sandbox) prepare() error {
+	var err error
+
+	err = s.PrepareTmpDir()
+	if err != nil {
+		return err
+	}
+
+	err = s.PrepareContainer()
 	if err != nil {
-		log.Fatal(err)
+		return err
+	}
+
+	return nil
+}
+
+// prepares the execution environment by copying all resources (user code, input files,
+// and execution payload) into a temporary directory.
+//
+func (s *sandbox) PrepareTmpDir() error {
+	// create tmp directory for keeping all code and inputs
+	tmpFolder, err := ioutil.TempDir(s.options.folder, TmpDirPrefix)
+	if err != nil {
+		return err
 	}
 
+	// modify perms on tmp dir
 	if err := os.Chmod(tmpFolder, 0777); err != nil {
-		log.Fatal(err)
+		return err
 	}
 
 	// record tmpdir for easy deletion
 	s.options.folder = tmpFolder
 
+	// copy the Payload dir into the tmp dir. for more details on what the
+	// Payload dir is, check out the TODO (cw|4.29.2018) fill this in...
 	err = s.copyPayload()
 	if err != nil {
-		log.Fatal(err)
+		return err
 	}
 
-	//log.Println(s.code)
+	// write source file into tmp dir
+	// TODO (cw|4.29.2018) we should be able to write an arbitrary number of files
```

Comment:
Makes sense! iow, consume a list of source files and map `WriteFile` over it?

Reply:
oops, i meant to take this comment out-- it was meant to say "write an arbitrary number of resource files" (i.e. files which the user submitted source can read and write to if it needs to load, say parameters or some data for the code to work). this has been addressed in the PR so i can take out this TODO.